### DISCUSSION TITLE
[BG] How to add bots to other BG including custom?

### DISCUSSION THREAD

**Discussion by artDeil4ek** [trickerer/Trinity-Bots#780] [Q&A]
Hello there.
I made a test BG map and i want to spawn your bots on it. Just adding records in creature_template_npcbot_wander_nodes doesn't work. And i've tried to add my BG to worldserver config file, but it causes some of errors.
My question is: How to add bots to BG properly?
Thank you.
![2024-04-17_09-26-32](https://github.com/trickerer/Trinity-Bots/assets/62905878/21c5d347-47fc-4ab8-93c6-904f0d56bdda)
![2024-04-17_09-25-17](https://github.com/trickerer/Trinity-Bots/assets/62905878/a72cf65c-0d3f-4a70-84f9-a2982e0b282a)
![2024-04-17_09-08-54](https://github.com/trickerer/Trinity-Bots/assets/62905878/6e8f05e5-7a84-405f-bc4f-c3137de9c361)

**Comment by trickerer** [comment]
Sorry but I doubt this is something you will be able to handle with no previous experience. It's a lengthy process.

First you need to create a proper wander node map with spawn nodes for both factions and objective nodes.
Then you need to implement npcbot handling in BG class (/src/game/Battelgrounds/Zones/). Battleground is made to only accept players so each part of BG logic which requires player class manipulation needs to be implemented for bots also. Take Warsong Gulch as an example. Open `Batteground.h`, `BattlegroundWS.h` and `BattlegroundWS.cpp` and search for `npcbot` comments.
Some things like area triggers cannot be handled from BG class since they rely on client sending proper opcode, so they have to be handled from botAI (`bot_ai::OnWanderNodeReached()`)
To script bots handling BG objectives and activating objects you again have to do that from within `bot_ai::OnWanderNodeReached()`
If wander node selection logic is complex and cannot be configured using node linking and flags alone, additional logic has to be added to `bot_ai::GetNextBGTravelNode()`

Also I'm a little curious to see the actual errors you get.

**Reply by artDeil4ek** [reply]
Hello. It's me again
now i'm getting this message:
![2024-05-21_10-46-32](https://github.com/trickerer/Trinity-Bots/assets/62905878/e89e267d-5482-4ffa-8eb1-4eca7f07e4b2)
any idea what am i doing wrong?

**Reply by trickerer** [reply]
You probably either have not enough free Alliance bots in 'bots of enabled classes' pool or there is no wander node in the map with maxlevel >= 55 available for Alliance team.

**Reply by artDeil4ek** [reply]
There's no problem with other BG's (AV,AB,WSG) so i think the problem it's wander nodes. I've added some nodes but i don't know how it works properly.
![2024-05-21_12-26-29](https://github.com/trickerer/Trinity-Bots/assets/62905878/8005c7a9-ad2d-4418-89a3-bafcba0b08a3)
What does mean this flags and links? And how can i set node for A/H?

**Reply by trickerer** [reply]
WP flag values can be found in `botwanderful.h`. You need at least one spawn for alliance (3) and one for horde (5), unless you want to make a spawn point for both teams somewhere.

Links are just a list of WPs this WP links to. This of it as a direction, in your screenshot 5088 links to 5089 or `5088->5089`: bots can move from 5088 to 5089, flags can partially restrict that.

**Comment by Conan1996-2** [comment]
I just came across this and love the way its explained. I may start working on Eye of the Storm, considering it's not any more complicated than Arathi Basin, only the flag would require additional programming along the lines of Warsong Gulch. Other than that, the logic is pretty much the same, capture bases.

**Comment by ilkashpilkerman** [comment]
Thanks for some explanation
![WoWScrnShot_081424_234122](https://github.com/user-attachments/assets/1185f132-8079-46bd-8448-3dd9f22dcd6c)
